Upgrade System RFD (v2)

fspmarshall opened the pull request:
Automatic upgrade system proposal, version 2 (re-written since the first version was mostly a poor attempt at doing what TUF already does).

The RFD is ready to merge, but will continue to be a 'living' document as work on this family of features progresses.

[Rendered](https://github.com/gravitational/teleport/blob/fspmarshall/rfd-upgrade-system-v2/rfd/XXXX-upgrade-system.md)

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,659 @@
+
+## Why
+
+Teleport must be perdiocially updated in order to integrate security patches.  Regular
```
```suggestion
Teleport must be periodically updated in order to integrate security patches.  Regular
```

espadolini commented on rfd/XXXX-upgrade-system.md:
```suggestion
- The installer accepts the installation `target` as an "argument", so that the version and any
```

espadolini reviewed (commented):
One concern that I have is that with graceful upgrades we might have two different versions of teleport concurrently running on a single node - i.e. both reporting with the same node_uuid and using the exact same process storage (and thus credentials). Would that fall into the bucket of "version unclear, try again later"? Would we require a mechanism to check the health of both old and new instances separately? As it's currently coded, I believe it's entirely random whether checking /healthz hits the old or the new instance.

klizhentas commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,659 @@
+needs to understand the `install-controller` will depend on the specific controller.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
```
I agree with the model of converging to a state by a pluggable controller. It is very close (modeled after?) Kubernetes controllers.

klizhentas commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,659 @@
+(arch/fips/etc) in addition to version. Make version inventory info available
+via `tctl`.
+
+2. MVP: Implement the `version-directive` resource, version reconciliation loop,
```
I agree on the MVP scope - and leveraging existing packages + TUF as much as possible

klizhentas commented on rfd/XXXX-upgrade-system.md:
Yes, interested on this as well. The node state should probably keep track of the history of runs and the results. 

We should also have a built-in strategy to avoid thundering herd outage. E.g. use default K8s deployment rollout/rollback strategy that should auto-stop if there is a percentage of nodes that did not start after the script run.

fspmarshall replied:
Added an 'Inventory Status and Visibility' section that discusses some basic state tracking, but there is definitely room for more.  Look it over and let me know what you think.

klizhentas commented on rfd/XXXX-upgrade-system.md:
On my personal/lab computers, I'd be happy to run maximum use-case. I think it could be as simple as having a controller that polls TUF and updates the internal state agressively. If it fails for me, no big deal, but at least I'd be always up to date on my homelab.

klizhentas commented on rfd/XXXX-upgrade-system.md:
Can you say a couple of words about future K8s controller as well? Afaik this should map perfectly on K8s controllers, but wanted to make sure.

klizhentas commented on rfd/XXXX-upgrade-system.md:
Can we have two other use-cases? 

* Dry run use-case is always helpful so it can tell you what will happen without actually running things.

Dry run use-case is also helpful to mitigate massive network spike. In dry-run use-case packages can be fetched locally on each machine, but not actually installed. This will help to avoid another thundering herd issue with 10K teleport nodes pulling the packages and restarting.

* Alert but don't upgrade use-case can set some internal "alert" state if your system is out of date. It is similar to dry run, but this can be turned on by default. Kind of like Chrome "green/yellow/red" button

To me, the biggest value of this system even before the upgrade would be to have a periodic job on teleport side that checks against TUF, does not upgrade anything, but sets internal alert in the dashboard or teleport status that tells that system is running vulnerable software. I'd like to include this in MVP - this will solve a lot of issues for many users who are not aware of the vulnerabilities today.

klizhentas commented on rfd/XXXX-upgrade-system.md:
We can steal the idea of rollingUpgradePolicy from K8s with a simple upgrade policy tracking the failure threshold and stopping at it.

klizhentas commented on rfd/XXXX-upgrade-system.md:
@reedloden would be cool for us to think through possible attacks on this model of upgrade. Can we mount an attack to keep Teleport on stale versions for example.

klizhentas commented on rfd/XXXX-upgrade-system.md:
We can make the check of the client automatic during the client ping. Clients can simply alert user to upgrade to the latest version. For restricted environments we can refuse clients of the older versions as well. But I think the upgrade should be triggered by users/endpoint managers.

Examples are `terraform` that checks its own version on every run

klizhentas commented on rfd/XXXX-upgrade-system.md:
Teleport has upgrade/rollback sequence - https://goteleport.com/docs/setup/admin/graceful-restarts/#graceful-upgrades we should leverage that

klizhentas commented on rfd/XXXX-upgrade-system.md:
version controller is better, as manifest is something very specific, not clear to most users.

klizhentas reviewed (commented):
I've done a first pass. I agree with the concept, but we need to zoom into specific details. Here's what I think this RFD should cover:

* Alerts (the check part, set the alert, but don't upgrade). This will deliver most value from the security perspective for most users with minimum crash risks.
* Client version checks with warnings/blocks. Same as the alerts, this will deliver most value for most users with minimum risks.
* Use-case with examples. We should polish UX by looking at administrator's and users actions (tctl commands, resources, error messages, rollbacks). This will show some flaws in the design.
* More refined implementation phases. There is a lot of scope, may be for 1-2 person years of work, so having discipline on the stages of the roll out will be critical. 
* Things that are out of scope.  If I understand correctly, this does not cover Auth servers. It's OK, but would be good to mention that auth servers are out of scope.

Edge cases that we should cover:

* Rollbacks (using signals?)

Nice to have: 

* Section that mentions future integration with K8s (will be relevant to cloud)

P.S. 

Talking about your question on certs, I agree that having one combined cert per process is probably better - easier to manage, rotate and reload.

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+System for automatic upgrades of teleport installations.
+
+
+## Why
```
Nit: I would mention Teleport Cloud in the why section, as it's a major driver for this feature (we can't upgrade proxy/auth on tenants where the nodes are out of date).

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+for key rotation, censorship detection, package validation, and much more.
+
+
+### High-Level Goals
```
Who is our target customer that we're building this for? The ones who run Teleport directly on VMs w/ systemd? Those who run in k8s with helm? Those who run our AMIs or docker images? None of these? All of them?

I think calling this out early will help readers understand the motivation behind the decisions we're making here. Plus, it's often best to solve a smaller problem really well than to do an okay job at solving a larger problem.

knisbet replied:
May also require some mention of different installation angles, such as if the customer uses RPM. If this is just a binary on disk update, there may be some interactions with packaging systems, like the version of the package not matching the binary, or if it will just call rpm/deb/etc to update.

fspmarshall commented:
@klizhentas I think I've addressed most of your feedback now.  Still need to add  tsh UX for alerts, but otherwise I think its all there.  A decent portion of the document has been expanded or entirely rewritten since your first pass, but the sections most relevant to your feedback are probably 'Inventory Status and Visibility', 'Version Directive Flow',  'Notification-Only Install Controller', and 'Rollbacks'.

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+controller would be any program with the permissions necessary to update the
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
```
Would it be more accurate to say that the version reconciliation loop runs in the version controller (as opposed to the auth server)?

fspmarshall replied:
No, the two things are different.  A version controller generates a detailed description of what the desired state is, and what installers to use to get to that state.  The controller could be anything... might run on the auth server, might be plugin, or might even be a human manually writing the directive.
The reconciliation loop is the builtin teleport functionality that orchestrates the details of how the directive gets applied.  E.g. if a version controller generates a directive to the affect of "all servers with label `env:prod` should be running `v1.2.3`, which should be installed using `local-script/my-installer`", the reconciliation loop is what looks at that directive and decides which individual servers actually need to have the installer run, then invokes the installer for those servers.

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+probably want an API that will support plugins that can emit alerts to external locations
+(e.g. a slack channel).
+
+In this usecase teleport is serving up recommendations based on external state, so a client
```
What is the source of truth for this external state? Github releases? Some new service?

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+Cluster administrators explicitly set the desired version(s) of teleport installations, and
+have a specific in-house installation script that should be used for upgrades. Teleport cluster
+may not even have any access to the public internet.  The installation target may be an internally
+built fork of teleport.
```
Supporting installation of forks may be an important feature, but I'm not sure I would describe it as a "minimal" use case. There's lots to consider there. Is this feature something we can defer?

fspmarshall replied:
Its minimal in the sense that teleport's responsibilities/involvement is minimal.  This scenario is about the case where users are just using teleport to trigger their own install mechanism that is totally unspecified from teleport's perspective. The user-supplied script does everything.  The script might install a fork... it also might just create a file named `plz-upgrade-soon.txt` on the host that the interns know to look for.

fspmarshall replied:
> There are only two hard things in Computer Science: cache invalidation and naming things.

Any suggestions?  I've become accustomed to us using terms like client/server when referring to the relationship between auth and non-auth instances, but I'm good with anything that evokes directionality.

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+
+Cluster administrators explicitly set the desired version(s) of teleport installations, and
+have a specific in-house installation script that should be used for upgrades. Teleport cluster
+may not even have any access to the public internet.  The installation target may be an internally
```
Where and how will teleport get information about new releases without access to the internet?

fspmarshall replied:
It doesn't. The point of this scenario is that cluster administrators are manually setting the desired state (i.e. creating their own `version-directive`).  I'll try to clarify it a bit.

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+compatibility are likely handled by the script.
+
+In this minimal usecase teleport's role is primarily as a coordinator.  It detects when and where user
+provided scripts should be run, and invokes them.
```
What happens between the time that teleport invokes the script and the time the upgrade completes? 

- Does teleport somehow know/track that it has an upgrade in progress?
- Is there a period after which time we consider the upgrade to have failed?
- What if the script fails? Do we retry? With any type of backoff?

zmb3 commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+all release artifacts (or at least the subset that can be built deterministically).
+
+- Verification Signer: A separate environment isolated from the main build system will independently build
+and all deterministic artifacts. All metadata will be independently generated and signed by this system.
```
build and what all artifacts? Is there a word missing here?

zmb3 reviewed (commented):
Only got halfway through, but wanted to submit this while it's fresh.

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+
+The proposed system consists of at least the following components:
+
+- `version-directive`: A static resource that describes the desired state of
+versioning across the cluster. The directive includes matchers which allow
+the auth server to match individual teleport instances with both the appropriate
+installation target, and the appropriate installation method. This resource may
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
```
Is this a singleton resource or can there be multiples of it? What sort of access control do we need for it? Do we need to send this around? Do we have to be careful about sending this around to proxies/nodes to avoid accidental disclosure of other resources in the cluster?

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+- The version directive resource (mostly complete already if we did the notification-only
+system first), and the version directive flow.
+
+- The `local-script` installer, and basic installer configuration API (get/put/del).
```
Will we provide apt-based and rpm-based update scripts in their respective packages?

fspmarshall replied:
Doable, I wouldn't want to mess with our package builds unless we really saw demand for it, and if we did it might be better to just have teleport come with "builtins" for specific package managers, that way we could do bug fixes/improvements behind the scenes.  To begin with, examples in the docs are probably simples/best.

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+normal install controller. Requires the new teleport installation to work at least well
+enough to perform any functions required by the install controller.
+
+2. Local rollback: The previous teleport installation remains cached during the upgrade, and some local process
+monitors the health of the new version. If the new version remains unhealthy for too long, it is forcibly
+terminated and the previous installation is replaced.
```
Is this planned to expand or replace the current readiness check for the new process during HUP?

espadolini reviewed (commented):
Left a few comments and a bunch of questions

fspmarshall commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
+
+- `version-controller`: A pluggable component responsible for generating the
```
Yes. The previous sentence says that directives can be manually created :D

Updating this section to better clarify that version controllers are optional.

fspmarshall commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+
+The proposed system consists of at least the following components:
+
+- `version-directive`: A static resource that describes the desired state of
+versioning across the cluster. The directive includes matchers which allow
+the auth server to match individual teleport instances with both the appropriate
+installation target, and the appropriate installation method. This resource may
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
```
Both!  There is a singleton "active" directive, and any number of draft and pending directives. Next push will update this section to include a link to the 'Version Directive Flow' section, which covers this.

Only auth servers, plugins, and `tctl` need to read the resource.  Nodes will see messages telling them what target to install and how to do it, but not what peers might also be affected.  Users with access to a specific server resource will see if that resource has a newer version available, or if there was a recent install attempt they will know what version and the name of the install method, but not the details (e.g. I could see that the `node.example.com` has status `installing v1.2.3 via local-script/apt`, but I can't see what the contents or inputs of the script are, or the details of the policy that resulted in that install attempt).

fspmarshall commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+}
+
+// ServerHello is the first message sent by the server. 
+message ServerHello {
```
It kind of does right now, but that's basically a coincidence.  I wanted the key information to all be part of the stream, since its possible that subsequent calls hit different auth servers.  Also, this message is over an authenticated connection, so while it does not currently contain any non-public info, it is able to include it in the future.

fspmarshall commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+    // Version is the currently running teleport version.
+    string Version = 1;
+    // ServerID is the unique ID of the server.
+    string ServerID = 2;
```
Unfortunately I think its the file whose nomenclature is out of date.  We now use EC2 instance IDs in some cases, so I think `ServerID` is becoming the more standard naming scheme.

espadolini replied:
Perhaps `HostID` then?

klizhentas commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+### Usage Scenarios
+
+
+#### Notification-Only Usecase
```
@xinding33 can we add some mockups of how the UI notifications will look like for users? Similar to the banner we have with "assume roles?"

klizhentas commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+In this case we want options for both displaying suggested version (or just a "needs upgrade"
+badge) on inventory views, and also probably some kind of cluster-wide alert that can be
+made difficult to ignore (e.g. a message on login or a banner in the web UI). We also
+probably want an API that will support plugins that can emit alerts to external locations
```
@fspmarshall I think adding a state to our prometheus endpoint and possibly `/secure` endpoint with `200` and non 200 endpoint for scripting will be super helfpul

klizhentas commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+to its connected nodes, allowing for real-time signaling.
+
+
+### Notification-Only System (?)
```
@fspmarshall @xinding33 @russjones I would suggest we enable this notification-only system by default for all OSS or at least have it as an opt-out during install because it's in user's benefit. 

We should add UI notifications and tsh command outputs every time users use an obsolete system. 

For example, in UI and CLI:

```
$ tsh ls
WARNING: you are using obsolete Teleport that has CVE .... Upgrade the system immediately.
```

Can we make this system to be enabled by default for OSS and have opt-out for enterprise (more sensitive to outbound network calls). Can we make this a part of the Notification-only system deliverable?

I think this will deliver maximum security value with minimum engineering efforts.

klizhentas reviewed (commented):
I have done a second pass:

* All fundamentals (TUF, level-based system, phases and basic CLI UX make sense to me). 

Here are the key comments:

* Can you please add an example section with imaginary documentation and CLI commands UX with various scenarios so we can imagine administrator flow better?

* For notification system, can we find a way to notify users and administrators in UI and CLI and opt-in all open source users for notifications and for enterprise users have this as a opt-out. @xinding33 can you take a look at how UI notifications may look like with design team? 

cc @russjones

P.S. Here are some thoughts on how notification system may look like:

@fspmarshall @xinding33 @russjones I would suggest we enable this notification-only system by default for all OSS or at least have it as an opt-out during install because it's in user's benefit. 

We should add UI notifications and tsh command outputs every time users use an obsolete system. 

For example, in UI and CLI:

```
$ tsh ls
WARNING: you are using obsolete Teleport that has CVE .... Upgrade the system immediately.
```

Can we make this system to be enabled by default for OSS and have opt-out for enterprise (more sensitive to outbound network calls). Can we make this a part of the Notification-only system deliverable?

I think this will deliver maximum security value with minimum engineering efforts.

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+    // Env is the script env variables.
+    map<string, string> Env = 2;
+    // Shell is the optional shell override.
+    string Shell = 3;
```
I was thinking of preexisting on-disk files, but inline scripts make sense too.

russjones commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1222 @@
+---
+
+# RFD XXXX - Upgrade System
+
```
```suggestion
# Required Approvers
* Engineering: @klizhentas && (@zmb3 || @rosstimothy || @espadolini)
* Product: (@klizhentas || @xinding33)
```

raisa-mase commented:
What is the process is for testing updates before they get pushed up to production?

raisa-mase commented:
What is the process is for testing updates before they get pushed up to production?

xinding33 commented:
@klizhentas I want to clarify a few comments from above since you used both the opt-out and opt-in terminologies. In particular:

> For notification system, can we find a way to notify users and administrators in UI and CLI and opt-in all open source users for notifications and for enterprise users have this as a opt-out.

> I would suggest we enable this notification-only system by default for all OSS or at least have it as an opt-out during install because it's in user's benefit.

Opt-out = enabled by default, user has to manually disable.
Opt-in = disabled by default, user has to manually enable.

I think we should make the notification-only system opt-out for everything (i.e. OSS, Cloud, and on-prem) since, as you mentioned above, it's in the user's benefit.

Why even give users the ability to opt-out? Because people do value customizability, especially as it pertains to notifications and other types of warnings (especially in open source products).

Can you clarify your suggestion?

phosphore commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1568 @@
+fork of teleport.
+
+In this case, we want to provide the means to specify the target version and desired script. Teleport
+should then be able to detect when an instance is not running the target that it aught to, and
```
augh -> ough

espadolini commented on rfd/XXXX-upgrade-system.md:
```diff
@@ -0,0 +1,1658 @@
+The trick to making sure a TUF-based system really lives up to the promise of the framework, is to have
+a good model for how the TUF metadata is generated and signed in the first place. This is where we come
+to the heart of our specific security model. We will leverage TUF's thresholded signature system and
+go's ability to produce deterministic builds in order to establish isolated cross-checks that can
```
How far are we from deterministic builds as of today?

espadolini commented on rfd/XXXX-upgrade-system.md:
How do we know if one is allowed to upgrade from one version to another?

fspmarshall replied:
Updated installer discussion to better clarify.  Version compatibility will check basically be "same major version or one major version newer".

espadolini commented on rfd/XXXX-upgrade-system.md:
Would said installer still get matched?

fspmarshall replied:
Checking expected target attributes is definitely something we *could* keep track of for matching.  My current feeling is that it might be better not to do this.  I'm wary of making the system too clever, as I want to keep it easy to review and reason about. Installer selection will definitely need to take node version into account since there will be an intermediate time period where folks may have some nodes that support `tuf` installers, but some that only support `script`. Additionally, one of the open questions is about whether or not to support nodes explicitly opting into a limited subset of the available installers (the most obvious choice being to explicitly only allow the use of the `tuf` installer, since it is likely to be the most secure once it is stabilized). If we did something like that, we'd definitely want it to be part of the matching logic as well.

espadolini commented on rfd/XXXX-upgrade-system.md:
Do we get a match on the directive even if there's no compatible installer on the node? If not, how can one choose to only receive upgrade notifications?

fspmarshall replied:
Yes.  If there are no compatible installers on the node, it still matches.  Good for notifications, as you mentioned.  Also important because we want incorrect configurations to fail rather than try to continue and possible create unexpected behaviors.

fspmarshall committed:
RFD 0090 - Upgrade System
